Standardize database progress output across all APIs

tanishadahale opened the pull request:
<!-- Please be sure to check out our developer guide,
https://radis.readthedocs.io/en/latest/dev/developer.html -->

### Description
<!-- Provide a general description of what your pull request does. -->

This pull request is to address to #868 which is solved in this PR by Standardizing database progress output across all APIs.
- Add DatabaseProgressPrinter utility class for consistent output formatting
- Add tqdm progress bars for downloads (HITEMP, ExoMol, HITRAN)
- Suppress HAPI verbose output (version banner, byte-by-byte progress)
- Standardize headers: 'MOLECULE - DATABASE VERSION - Action'
- Update all database APIs: HITEMP, ExoMol, HITRAN, GEISA

Resolves issue about inconsistent progress display across database fetchers



<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->

```
--------------------------------------------------------------------------------
CO2 - HITEMP 2024 - Downloading and processing 1 chunks for range 3500.0-3505.0 cm⁻¹
--------------------------------------------------------------------------------

Download:
- Download 1 file(s) missing out of 1.
Found target in compressed blocks: 30009 to 30519
Wavenumber range: 3468.688960 to 3508.820030 cm-1
Downloading: 100%|█████████████████████████████████████████████████████████████████████| 74.2M/74.2M [01:08<00:00, 1.14MB/s]
Chunk Download complete
Written decompressed data to /Users/darshanpatil/.radisdb/hitemp/CO2_HITEMP2024/CO2_02_03468-03508_HITEMP2024.par
Processing chunks: 100%|█████████████████████████████████████████████████████████████| 1/1 [01:29<00:00, 89.16s/it, parsing]
- Combining parsed data from all chunks...
HAPI version: 1.2.2.4
```
```
--------------------------------------------------------------------------------
CO2 - HITEMP - Downloading database
--------------------------------------------------------------------------------

Download:
- Downloading 02_03500-03750_HITEMP2010.zip (1/1)
Login successful.
Downloading: 100%|█████████████████████████████████████████████████████████████████████| 37.3M/37.3M [00:06<00:00, 6.19MB/s]
Download complete. Parsing CO2 database to /Users/darshanpatil/.radisdb/hitemp/CO2-02_03500-03750_HITEMP2010.h5
The parsing/conversion is usually very fast (e.g., HITEMP OH takes only a few seconds) but can be slightly longer in some cases (e.g., a single HITEMP 2010 CO₂ file takes about 1 minute).
Added HITEMP-CO2-2010 database in /Users/darshanpatil/radis.json
Keep only relevant input file: ['/Users/darshanpatil/.radisdb/hitemp/CO2-02_03500-03750_HITEMP2010.h5']
Calculating Equilibrium Spectrum
Physical Conditions
```
```
/Users/darshanpatil/Documents/radis/radis/misc/warning.py:427: AccuracyWarning: Using lines from exomol but partition functions from hapifor consistency we recommend using lines and partition functions from the same database
  warnings.warn(WarningType(message))
--------------------------------------------------------------------------------
CO2 - EXOMOL UCL-4000 - Fetching 12C-16O2
--------------------------------------------------------------------------------
  - Total download size: 0.000010 GB
  - Downloading 12C-16O2__UCL-4000.def
    12C-16O2__UCL-4000.def: 100%|██████████████████████████████████████████████████████| 10.4k/10.4k [00:00<00:00, 56.3kB/s]
  - Total download size: 0.000116 GB
  - Downloading 12C-16O2__UCL-4000.pf
    12C-16O2__UCL-4000.pf: 100%|██████████████████████████████████████████████████████████| 122k/122k [00:00<00:00, 168kB/s]
  - Total download size: 0.068052 GB
  - Downloading 12C-16O2__UCL-4000.states.bz2
    12C-16O2__UCL-4000.states.bz2: 100%|███████████████████████████████████████████████| 69.7M/69.7M [00:10<00:00, 6.76MB/s]
  - Total download size: 0.000036 GB
  - Downloading 12C-16O2__H2.broad
    12C-16O2__H2.broad: 100%|██████████████████████████████████████████████████████████| 4.41k/4.41k [00:00<00:00, 21.7MB/s]
  - Downloading 12C-16O2__He.broad
    12C-16O2__He.broad: 100%|██████████████████████████████████████████████████████████| 4.41k/4.41k [00:00<00:00, 19.7MB/s]
  - Downloading 12C-16O2__air.broad
    12C-16O2__air.broad: 100%|█████████████████████████████████████████████████████████| 4.41k/4.41k [00:00<00:00, 11.8MB/s]
  - Downloading 12C-16O2__self.broad
    12C-16O2__self.broad: 100%|████████████████████████████████████████████████████████| 8.47k/8.47k [00:00<00:00, 1.28MB/s]
  - Downloading 12C-16O2__Ar.broad
Warning: Couldn't download .broad file
  - Downloading 12C-16O2__CH4.broad
Warning: Couldn't download .broad file
  - Downloading 12C-16O2__CO.broad
Warning: Couldn't download .broad file
  - Downloading 12C-16O2__CO2.broad
Warning: Couldn't download .broad file
  - Downloading 12C-16O2__H2.broad
    12C-16O2__H2.broad: 100%|██████████████████████████████████████████████████████████| 4.41k/4.41k [00:00<00:00, 18.7MB/s]
  - Downloading 12C-16O2__H2O.broad
    12C-16O2__H2O.broad: 100%|█████████████████████████████████████████████████████████| 11.1k/11.1k [00:00<00:00, 35.3kB/s]
  - Downloading 12C-16O2__N2.broad
Warning: Couldn't download .broad file
  - Downloading 12C-16O2__NH3.broad
Warning: Couldn't download .broad file
  - Downloading 12C-16O2__NO.broad
Warning: Couldn't download .broad file
  - Downloading 12C-16O2__O2.broad
Warning: Couldn't download .broad file
  - Downloading 12C-16O2__NH3.broad
Warning: Couldn't download .broad file
  - Downloading 12C-16O2__CS.broad
Warning: Couldn't download .broad file
  - Broadening files: 6 succeeded, 10 failed
  - Caching states data to pytables format. This will be faster on subsequent runs.
--------------------------------------------------------------------------------
CO2 - EXOMOL UCL-4000 - Loading 12C-16O2
--------------------------------------------------------------------------------
- Local folder: /Users/darshanpatil/.radisdb/exomol/CO2/12C-16O2/UCL-4000

Transition files:
- File 12C-16O2__UCL-4000__03000-04000.trans
  - Total download size: 0.501758 GB
  - Downloading 12C-16O2__UCL-4000__03000-04000.trans.bz2
    12C-16O2__UCL-4000__03000-0400: 100%|████████████████████████████████████████████████| 514M/514M [01:29<00:00, 6.02MB/s]
  - Caching the *.trans.bz2 file to the pytables (*.h5) format. After the second time, it will become much faster.
Broadening code level: m0
Broadening code level: None
/Users/darshanpatil/Documents/radis/radis/api/exomolapi.py:1613: UserWarning: The broadening file contains this broadening code: None. This broadening code is NOT implemented yet.
Using default parameters instead.
  warnings.warn(

ExoMol database loading complete
```
```
--------------------------------------------------------------------------------
CO2 - HITRAN - Downloading database
--------------------------------------------------------------------------------

Download:
- Downloading all isotopes for CO2
HAPI version: 1.2.2.4
To get the most up-to-date version please check http://hitran.org/hapi
ATTENTION: Python versions of partition sums from TIPS-2021 are now available in HAPI code

           MIT license: Copyright 2021 HITRAN team, see more at http://hitran.org. 

           If you use HAPI in your research or software development,
           please cite it using the following reference:
           R.V. Kochanov, I.E. Gordon, L.S. Rothman, P. Wcislo, C. Hill, J.S. Wilzewski,
           HITRAN Application Programming Interface (HAPI): A comprehensive approach
           to working with spectroscopic data, J. Quant. Spectrosc. Radiat. Transfer 177, 15-30 (2016)
           DOI: 10.1016/j.jqsrt.2016.03.005

           ATTENTION: This is the core version of the HITRAN Application Programming Interface.
                      For more efficient implementation of the absorption coefficient routine, 
                      as well as for new profiles, parameters and other functional,
                      please consider using HAPI2 extension library.
                      HAPI2 package is available at http://github.com/hitranonline/hapi2

Downloading isotopes: 100%|███████████████████████████████████████████████████████████████████| 9/9 [01:08<00:00,  7.66s/it]

HITRAN database download complete
Added HITRAN-CO2 database in /Users/darshanpatil/radis.json
```
```
--------------------------------------------------------------------------------
CO2 - GEISA - Downloading database
--------------------------------------------------------------------------------

Download:
- Downloading line_GEISA2020_asc_gs08_v1.0_co2 (1/1)
Downloading: 100%|███████████████████████████████████████████████████████████████████████| 128M/128M [00:20<00:00, 6.66MB/s]
Successfully downloaded https://aeris-geisa.ipsl.fr/geisa_files/2020/Lines/line_GEISA2020_asc_gs08_v1.0_co2
Added GEISA-CO2 database in /Users/darshanpatil/radis.json
```

Fixes #868

tanishadahale committed:
feat: Standardize database progress output across all APIs

- Add DatabaseProgressPrinter utility class for consistent output formatting
- Add tqdm progress bars for downloads (HITEMP, ExoMol, HITRAN)
- Suppress HAPI verbose output (version banner, byte-by-byte progress)
- Standardize headers: 'MOLECULE - DATABASE VERSION - Action'
- Update all database APIs: HITEMP, ExoMol, HITRAN, GEISA

Resolves issue about inconsistent progress display across database fetchers

tanishadahale committed:
Update author in database_progress.py

tanishadahale committed:
style: Apply linting fixes (black, isort, flake8)

tanishadahale commented:
Hey @minouHub could you please review this PR and approve remaining workflows?

abhinav-1305 commented:
Hi @tanishadahale , there are few test cases which are failing, pls check them 

```
FAILED radis/test/spectrum/test_spectrum.py::test_plot_by_parts - _tkinter.TclError: Can't find a usable tk.tcl in the following directories: 
    C:/hostedtoolcache/windows/Python/3.10.11/x64/tcl/tcl8.6/tk8.6 C:/hostedtoolcache/windows/Python/3.10.11/x64/tcl/tk8.6 C:/hostedtoolcache/windows/Python/3.10.11/lib/tk8.6 C:/hostedtoolcache/windows/Python/3.10.11/lib/tk8.6 C:/hostedtoolcache/windows/Python/lib/tk8.6 C:/hostedtoolcache/windows/Python/3.10.11/library

C:/hostedtoolcache/windows/Python/3.10.11/x64/tcl/tk8.6/tk.tcl: couldn't read file "C:/hostedtoolcache/windows/Python/3.10.11/x64/tcl/tk8.6/tk.tcl": no such file or directory
couldn't read file "C:/hostedtoolcache/windows/Python/3.10.11/x64/tcl/tk8.6/tk.tcl": no such file or directory
    while executing
"source C:/hostedtoolcache/windows/Python/3.10.11/x64/tcl/tk8.6/tk.tcl"
    ("uplevel" body line 1)
    invoked from within
"uplevel #0 [list source $file]"


This probably means that tk wasn't installed properly.
======== 1 failed, 129 passed, 143 deselected, 770 warnings in 52.35s =========
Error: Process completed with exit code 1.
```

minouHub commented:
Quick feedback (longer review later). These failing tests are acceptable. This is due to Windows CI not handling properly the plotting.

minouHub reviewed (changes requested):
Thank you for your hard work!

**A few comments:**

- [ ] "Suppress HAPI verbose output" 
I think that @erwanp had problems with that before. @erwanp could you confirm? I would keep it for the moment.

- [ ] Subsections 
There is a "download" subsection which is good. However this is the only one. Could you organize the different steps in several subsections? A `parsing` subection? A `broadening` subsection for ExoMol, with a summary to replace the `- Broadening files: 1 succeeded, 15 failed` coming out of the blue.

- [ ] ExoMol print is decomposed in two sections "MYTHOS" and "default". It should be a single one "MYTHOS (default)". Obviously the `default` should not be there if it's not the default database
- [ ] ExoMol running a 2nd time should not output anything because there is no download. Note that the warnings are ok.
- [ ] Titles are not consistent. In my opinion, "OH - HITRAN" is sufficient. No need to say "donwloading database" as it is printed after.
- [ ] There is a `s]` in the HITEMP `-----`
- [ ] `verbose=0` Make sure there is no output with verbose = 0, except for the warnings obviously

**My test:**
```
from radis import SpectrumFactory

sf = SpectrumFactory(10, 100,
                     molecule='OH',
                     isotope='1',
                     medium='vacuum',
                     verbose=1,    # more for more details
                     )
sf.fetch_databank('hitran')  
sf.fetch_databank('hitemp')  
sf.fetch_databank('exomol')   
```

**My output:**
```
--------------------------------------------------------------------------------
OH - HITRAN - Downloading database
--------------------------------------------------------------------------------

Download:
- Downloading all isotopes for OH
Downloading isotopes: 100%|██████████| 9/9 [00:03<00:00,  2.26it/s]

HITRAN database download complete
Added HITRAN-OH database in C:\Users\Nicolas Minesi\radis.json
Login successful.
Downloading: 100%|██████████| 900k/900k [00:00<00:00, 2.14MB/s]--------------------------------------------------------------------------------
OH - HITEMP - Downloading database
--------------------------------------------------------------------------------

Download:
- Downloading 13_HITEMP2020.par.bz2 (1/1)

Login successful.
Downloading: 100%|██████████| 900k/900k [00:00<00:00, 2.15MB/s]
Download complete. Parsing OH database to C:\Users\Nicolas Minesi\.radisdb\hitemp\OH-13_HITEMP2020.hdf5
The parsing/conversion is usually very fast (e.g., HITEMP OH takes only a few seconds) but can be slightly longer in some cases (e.g., a single HITEMP 2010 CO₂ file takes about 1 minute).
Added HITEMP-OH database in C:\Users\Nicolas Minesi\radis.json
Using the recommended ExoMol databases for OH
--------------------------------------------------------------------------------
OH - EXOMOL default - Fetching 16O-1H
--------------------------------------------------------------------------------
C:\Users\Nicolas Minesi\Python\radis\misc\warning.py:427: AccuracyWarning: Using lines from exomol but partition functions from hapifor consistency we recommend using lines and partition functions from the same database
  warnings.warn(WarningType(message))
  - Total download size: 0.000007 GB
  - Downloading 16O-1H__MYTHOS.def
    16O-1H__MYTHOS.def: 100%|██████████| 7.38k/7.38k [00:00<?, ?B/s]  - Total download size: 0.000233 GB
  - Downloading 16O-1H__MYTHOS.pf

    16O-1H__MYTHOS.pf: 100%|██████████| 244k/244k [00:00<00:00, 5.90MB/s]  - Total download size: 0.000214 GB
  - Downloading 16O-1H__MYTHOS.states.bz2

    16O-1H__MYTHOS.states.bz2: 100%|██████████| 224k/224k [00:00<00:00, 6.03MB/s]
  - Total download size: 0.000004 GB
  - Downloading 16O-1H__H2.broad
Warning: Couldn't download .broad file
  - Downloading 16O-1H__He.broad
Warning: Couldn't download .broad file
  - Downloading 16O-1H__air.broad
    16O-1H__air.broad: 100%|██████████| 3.92k/3.92k [00:00<?, ?B/s]  - Downloading 16O-1H__self.broad
Warning: Couldn't download .broad file
  - Downloading 16O-1H__Ar.broad
Warning: Couldn't download .broad file
  - Downloading 16O-1H__CH4.broad
Warning: Couldn't download .broad file
  - Downloading 16O-1H__CO.broad
Warning: Couldn't download .broad file

  - Downloading 16O-1H__CO2.broad
Warning: Couldn't download .broad file
  - Downloading 16O-1H__H2.broad
Warning: Couldn't download .broad file
  - Downloading 16O-1H__H2O.broad
Warning: Couldn't download .broad file
  - Downloading 16O-1H__N2.broad
Warning: Couldn't download .broad file
  - Downloading 16O-1H__NH3.broad
Warning: Couldn't download .broad file
  - Downloading 16O-1H__NO.broad
Warning: Couldn't download .broad file
  - Downloading 16O-1H__O2.broad
Warning: Couldn't download .broad file
  - Downloading 16O-1H__NH3.broad
Warning: Couldn't download .broad file
  - Downloading 16O-1H__CS.broad
Warning: Couldn't download .broad file
  - Broadening files: 1 succeeded, 15 failed
  - Caching states data to vaex format. This will be faster on subsequent runs.
--------------------------------------------------------------------------------
OH - EXOMOL MYTHOS - Loading 16O-1H
--------------------------------------------------------------------------------
- Local folder: C:\Users\Nicolas Minesi\.radisdb\exomol\OH\16O-1H\MYTHOS

Transition files:
- File 16O-1H__MYTHOS.trans
  - Total download size: 0.018387 GB
  - Downloading 16O-1H__MYTHOS.trans.bz2
    16O-1H__MYTHOS.trans.bz2: 100%|██████████| 18.8M/18.8M [00:00<00:00, 81.3MB/s]  - Caching the *.trans.bz2 file to the vaex (*.h5) format. After the second time, it will become much faster.

Broadening code level: m0
The default broadening parameters are used.

ExoMol database loading complete
C:\Users\Nicolas Minesi\Python\radis\api\exomolapi.py:1634: UserWarning: Could not load `16O-1H__self.broad`. The default broadening parameters are used.

  warnings.warn(
```
Running a second time:
```
Using the recommended ExoMol databases for OH
--------------------------------------------------------------------------------
OH - EXOMOL default - Fetching 16O-1H
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------
C:\Users\Nicolas Minesi\Python\radis\misc\warning.py:427: AccuracyWarning: Using lines from exomol but partition functions from hapifor consistency we recommend using lines and partition functions from the same database
  warnings.warn(WarningType(message))
OH - EXOMOL MYTHOS - Loading 16O-1H
--------------------------------------------------------------------------------
- Local folder: C:\Users\Nicolas Minesi\.radisdb\exomol\OH\16O-1H\MYTHOS

Transition files:
- File 16O-1H__MYTHOS.trans
Broadening code level: m0
The default broadening parameters are used.

ExoMol database loading complete
C:\Users\Nicolas Minesi\Python\radis\api\exomolapi.py:1634: UserWarning: Could not load `16O-1H__self.broad`. The default broadening parameters are used.

  warnings.warn(
```

This is a very nice work. I invite @HajimeKawahara, @dcmvdbekerom and @erwanp to provide feedback (even if incomplete) if they want.

tanishadahale committed:
Address PR #899 review: Standardize database output, fix headers, silence second run

tanishadahale commented:
Hello @minouHub I have Implemented all the requested changes, could you please review the changes, let me know if you need anymore changes, Thanks!

```
from radis import SpectrumFactory, plot_diff

sf = SpectrumFactory(3500, 3505,         # cm-1
                     isotope='1',
                     molecule='CO2',
                     verbose=1,    # more for more details
                     )

sf.fetch_databank("hitemp", "2024")
s1 = sf.eq_spectrum(2000, name = "HITEMP2024")

sf.fetch_databank("hitemp", "2010")
s2 = sf.eq_spectrum(2000, name = "HITEMP2010")

sf.fetch_databank("exomol")
s3 = sf.eq_spectrum(2000, name = "ExoMol")

sf.fetch_databank("hitran")
s4 = sf.eq_spectrum(2000, name = "hitran")

sf.fetch_databank("geisa")
s5 = sf.eq_spectrum(2000, name = "geisa")

```

```

--------------------------------------------------------------------------------
CO2 - HITEMP 2024 - Downloading and processing 1 chunks for range 3500.0-3505.0 cm⁻¹
--------------------------------------------------------------------------------

Download:
- Download 1 file(s) missing out of 1.
Found target in compressed blocks: 30009 to 30519
Wavenumber range: 3468.688960 to 3508.820030 cm-1
Downloading: 100%|███████████████████████████████████████████████████████████| 74.2M/74.2M [00:12<00:00, 6.38MB/s]
Chunk Download complete
Written decompressed data to C:\Users\Tanisha/.radisdb\hitemp\CO2_HITEMP2024\CO2_02_03468-03508_HITEMP2024.par
Processing chunks: 100%|██████████████████████████████████████████████████| 1/1 [02:07<00:00, 127.05s/it, parsing]
- Combining parsed data from all chunks...
HAPI version: 1.2.2.4
To get the most up-to-date version please check http://hitran.org/hapi
ATTENTION: Python versions of partition sums from TIPS-2021 are now available in HAPI code

           MIT license: Copyright 2021 HITRAN team, see more at http://hitran.org.

           If you use HAPI in your research or software development,
           please cite it using the following reference:
           R.V. Kochanov, I.E. Gordon, L.S. Rothman, P. Wcislo, C. Hill, J.S. Wilzewski,
           HITRAN Application Programming Interface (HAPI): A comprehensive approach
           to working with spectroscopic data, J. Quant. Spectrosc. Radiat. Transfer 177, 15-30 (2016)
           DOI: 10.1016/j.jqsrt.2016.03.005

           ATTENTION: This is the core version of the HITRAN Application Programming Interface.
                      For more efficient implementation of the absorption coefficient routine,
                      as well as for new profiles, parameters and other functional,
                      please consider using HAPI2 extension library.
                      HAPI2 package is available at http://github.com/hitranonline/hapi2

Calculating Equilibrium Spectrum
Physical Conditions
----------------------------------------
   Tgas                 2000 K
   isotope              1
   mole_fraction        1
   path_length          1 cm
   pressure             1.01325 bar
   self_absorption      True
   species              CO2
   state                X
   wavenum_max          3505.0000 cm-1
   wavenum_min          3500.0000 cm-1
Computation Parameters
----------------------------------------
   Tref                 296 K
   add_at_used          numpy
   broadening_method    voigt
   cutoff               0 cm-1/(#.cm-2)
   dbformat             hitemp-radisdb
   dbpath               C:\Users\Tanisha/.radisdb\hitemp\CO2_HITEMP2024\CO2_02_03468-03508_HITEMP2024.par
   diluent              air
   folding_thresh       1e-06
   include_neighbouring_lines  True
   isatom               False
   isneutral            None
   lbfunc               None
   memory_mapping_engine  auto
   neighbour_lines      0 cm-1
   optimization         simple
   parfuncfmt           hapi
   parsum_mode          full summation
   pfsource             default
   potential_lowering   None
   pseudo_continuum_threshold  0
   sparse_ldm           False
   truncation           50 cm-1
   waveunit             cm-1
   wstep                0.01 cm-1
   zero_padding         502
----------------------------------------
0.76s - Spectrum calculated
```
```
--------------------------------------------------------------------------------
CO2 - HITEMP
--------------------------------------------------------------------------------

Download:
- Downloading 02_03500-03750_HITEMP2010.zip (1/1)
Login successful.
Downloading: 100%|███████████████████████████████████████████████████████████| 37.3M/37.3M [00:10<00:00, 3.58MB/s]
Download complete. Parsing CO2 database to C:\Users\Tanisha\.radisdb\hitemp\CO2-02_03500-03750_HITEMP2010.h5
The parsing/conversion is usually very fast (e.g., HITEMP OH takes only a few seconds) but can be slightly longer in some cases (e.g., a single HITEMP 2010 CO₂ file takes about 1 minute).
Added HITEMP-CO2-2010 database in C:\Users\Tanisha\radis.json
Keep only relevant input file: ['C:\\Users\\Tanisha\\.radisdb\\hitemp\\CO2-02_03500-03750_HITEMP2010.h5']
Calculating Equilibrium Spectrum
Physical Conditions
----------------------------------------
   Tgas                 2000 K
   isotope              1
   mole_fraction        1
   path_length          1 cm
   pressure             1.01325 bar
   self_absorption      True
   species              CO2
   state                X
   wavenum_max          3505.0000 cm-1
   wavenum_min          3500.0000 cm-1
Computation Parameters
----------------------------------------
   Tref                 296 K
   add_at_used          numpy
   broadening_method    voigt
   cutoff               0 cm-1/(#.cm-2)
   dbformat             hitemp-radisdb
   dbpath               C:\Users\Tanisha\.radisdb\hitemp\CO2-02_03500-03750_HITEMP2010.h5
   diluent              air
   folding_thresh       1e-06
   include_neighbouring_lines  True
   isatom               False
   isneutral            None
   lbfunc               None
   memory_mapping_engine  auto
   neighbour_lines      0 cm-1
   optimization         simple
   parfuncfmt           hapi
   parsum_mode          full summation
   pfsource             default
   potential_lowering   None
   pseudo_continuum_threshold  0
   sparse_ldm           False
   truncation           50 cm-1
   waveunit             cm-1
   wstep                0.01 cm-1
   zero_padding         502
----------------------------------------
0.08s - Spectrum calculated
```
```
(base) PS C:\Users\Tanisha\Downloads\RADIS\radis> & C:/Users/Tanisha/AppData/Local/Programs/Python/Python313/python.exe c:/Users/Tanisha/Downloads/RADIS/radis/test.py
c:\Users\Tanisha\Downloads\RADIS\radis\radis\misc\warning.py:427: AccuracyWarning: Using lines from exomol but partition functions from hapifor consistency we recommend using lines and partition functions from the same database
  warnings.warn(WarningType(message))
Using the recommended ExoMol databases for CO2
  - Total download size: 0.000010 GB
  - Downloading 12C-16O2__Dozen.def
    12C-16O2__Dozen.def: 100%|████████████████████████████████████████████████████| 10.2k/10.2k [00:00<00:00, 44.7kB/s]
  - Total download size: 0.000116 GB
  - Downloading 12C-16O2__Dozen.pf
    12C-16O2__Dozen.pf: 100%|████████████████████████████████████████████████████████| 122k/122k [00:00<00:00, 165kB/s]
  - Total download size: 0.071626 GB
  - Downloading 12C-16O2__Dozen.states.bz2
    12C-16O2__Dozen.states.bz2: 100%|█████████████████████████████████████████████| 73.3M/73.3M [00:09<00:00, 7.83MB/s]
  - Total download size: 0.000036 GB

Broadening:
  - Downloading 12C-16O2__H2.broad
    12C-16O2__H2.broad: 100%|█████████████████████████████████████████████████████| 4.41k/4.41k [00:00<00:00, 13.6MB/s]
  - Downloading 12C-16O2__He.broad
    12C-16O2__He.broad: 100%|█████████████████████████████████████████████████████| 4.41k/4.41k [00:00<00:00, 11.4MB/s]
  - Downloading 12C-16O2__air.broad
    12C-16O2__air.broad: 100%|████████████████████████████████████████████████████| 4.41k/4.41k [00:00<00:00, 12.4MB/s]
  - Downloading 12C-16O2__self.broad
    12C-16O2__self.broad: 100%|████████████████████████████████████████████████████| 8.47k/8.47k [00:00<00:00, 362kB/s]
  - Downloading 12C-16O2__Ar.broad
Warning: Couldn't download .broad file
  - Downloading 12C-16O2__CH4.broad
Warning: Couldn't download .broad file
  - Downloading 12C-16O2__CO.broad
Warning: Couldn't download .broad file
  - Downloading 12C-16O2__CO2.broad
Warning: Couldn't download .broad file
  - Downloading 12C-16O2__H2.broad
    12C-16O2__H2.broad: 100%|█████████████████████████████████████████████████████| 4.41k/4.41k [00:00<00:00, 14.1MB/s]
  - Downloading 12C-16O2__H2O.broad
    12C-16O2__H2O.broad: 100%|█████████████████████████████████████████████████████| 11.1k/11.1k [00:00<00:00, 234kB/s]
  - Downloading 12C-16O2__N2.broad
Warning: Couldn't download .broad file
  - Downloading 12C-16O2__NH3.broad
Warning: Couldn't download .broad file
  - Downloading 12C-16O2__NO.broad
Warning: Couldn't download .broad file
  - Downloading 12C-16O2__O2.broad
Warning: Couldn't download .broad file
  - Downloading 12C-16O2__NH3.broad
Warning: Couldn't download .broad file
  - Downloading 12C-16O2__CS.broad
Warning: Couldn't download .broad file
  - Downloaded 6 broadening file(s): H2, He, air, self, H2, H2O
  - 10 broadening file(s) not available (using default parameters)
  - 10 broadening file(s) not available (using default parameters)
  - Caching states data to pytables format. This will be faster on subsequent runs.
--------------------------------------------------------------------------------
CO2 - EXOMOL - Fetching 12C-16O2 - Dozen (default)
--------------------------------------------------------------------------------
- Local folder: C:\Users\Tanisha\.radisdb\exomol\CO2\12C-16O2\Dozen

Download:
- File 12C-16O2__Dozen__03000-04000.trans
  - Total download size: 0.539887 GB
  - Downloading 12C-16O2__Dozen__03000-04000.trans.bz2
    12C-16O2__Dozen__03000-04000.t:  93%|███████████████████████████████████████▊   | 512M/553M [01:33    12C-16O2__Dozen__03000-04000.t:  93%|███████████████████████████████████████▉   | 514M/553M [01:34    12C-16O2__Dozen__03000-04000.t:  93%|████████████████████████████████████████   | 515M/553M [01:34    12C-16O2__Dozen__03000-04000.t:  93%|████████████████████████████████████████▏  | 517M/553M [01:34    12C-16O2__Dozen__03000-04000.t:  94%|████████████████████████████████████████▎  | 518M/553M [01:34    12C-16O2__Dozen__03000-04000.t:  94%|████████████████████████████████████████▍  | 519M/553M [01:34    12C-16O2__Dozen__03000-04000.t:  94%|████████████████████████████████████████▍  | 520M/553M [01:34    12C-16O2__Dozen__03000-04000.t:  94%|████████████████████████████████████████▌  | 521M/553M [01:34    12C-16O2__Dozen__03000-04000.t:  94%|████████████████████████████████████████▌  | 522M/553M [01:35    12C-16O2__Dozen__03000-04000.t:  95%|████████████████████████████████████████▋  | 523M/553M [01:35    12C-16O2__Dozen__03000-04000.t:  95%|████████████████████████████████████████▊  | 524M/553M [01:35    12C-16O2__Dozen__03000-04000.t:  95%|████████████████████████████████████████▊  | 525M/553M [01:35    12C-16O2__Dozen__03000-04000.t:  95%|████████████████████████████████████████▉  | 527M/553M [01:35    12C-16O2__Dozen__03000-04000.t:  96%|█████████████████████████████████████████  | 528M/553M [01:35    12C-16O2__Dozen__03000-04000.t:  96%|█████████████████████████████████████████▏ | 529M/553M [01:35    12C-16O2__Dozen__03000-04000.t:  96%|█████████████████████████████████████████▎ | 531M/553M [01:36    12C-16O2__Dozen__03000-04000.t:  96%|█████████████████████████████████████████▍ | 532M/553M [01:36    12C-16O2__Dozen__03000-04000.t:  96%|█████████████████████████████████████████▍ | 533M/553M [01:36    12C-16O2__Dozen__03000-04000.t:  97%|█████████████████████████████████████████▌ | 534M/553M [01:36    12C-16O2__Dozen__03000-04000.t:  97%|█████████████████████████████████████████▋ | 536M/553M [01:36    12C-16O2__Dozen__03000-04000.t:  97%|█████████████████████████████████████████▊ | 537M/553M [01:36    12C-16O2__Dozen__03000-04000.t:  97%|█████████████████████████████████████████▉ | 538M/553M [01:36    12C-16O2__Dozen__03000-04000.t:  98%|█████████████████████████████████████████▉ | 539M/553M [01:37    12C-16O2__Dozen__03000-04000.t:  98%|██████████████████████████████████████████ | 540M/553M [01:37    12C-16O2__Dozen__03000-04000.t:  99%|██████████████████████████████████████████▍| 545M/553M [01:38    12C-16O2__Dozen__03000-04000.t:  99%|██████████████████████████████████████████▌| 547M/553M [01:38    12C-16O2__Dozen__03000-04000.t:  99%|██████████████████████████████████████████▌| 548M/553M [01:38    12C-16O2__Dozen__03000-04000.t:  99%|██████████████████████████████████████████▋| 549M/553M [01:38    12C-16O2__Dozen__03000-04000.t:  99%|██████████████████████████████████████████▍| 545M/553M [01:38    12C-16O2__Dozen__03000-04000.t:  99%|██████████████████████████████████████████▌| 547M/553M [01:38    12C-16O2__Dozen__03000-04000.t:  99%|██████████████████████████████████████████▌| 548M/553M [01:38    12C-16O2__Dozen__03000-04000.t:  99%|██████████████████████████████████████████▋| 549M/553M [01:38    12C-16O2__Dozen__03000-04000.t:  99%|██████████████████████████████████████████▋| 550M/553M [01:38    12C-16O2__Dozen__03000-04000.t: 100%|██████████████████████████████████████████▊| 550M/553M [01:39    12C-16O2__Dozen__03000-04000.t: 100%|██████████████████████████████████████████▊| 551M/553M [01:39    12C-16O2__Dozen__03000-04000.t: 100%|██████████████████████████████████████████▉| 552M/553M [01:39    12C-16O2__Dozen__03000-04000.t: 100%|███████████████████████████████████████████| 553M/553M [01:39<00:00, 5.83MB/s]
  - Caching the *.trans.bz2 file to the pytables (*.h5) format. After the second time, it will become much faster.
Broadening code level: m0
Broadening code level: None
c:\Users\Tanisha\Downloads\RADIS\radis\radis\api\exomolapi.py:1652: UserWarning: The broadening file contains this broadening code: None. This broadening code is NOT implemented yet.
Using default parameters instead.
  )

ExoMol database loading complete
```
```
--------------------------------------------------------------------------------
CO2 - HITRAN - Downloading database
--------------------------------------------------------------------------------

Download:
- Downloading all isotopes for CO2
HAPI version: 1.2.2.4
To get the most up-to-date version please check http://hitran.org/hapi
ATTENTION: Python versions of partition sums from TIPS-2021 are now available in HAPI code

           MIT license: Copyright 2021 HITRAN team, see more at http://hitran.org. 

           If you use HAPI in your research or software development,
           please cite it using the following reference:
           R.V. Kochanov, I.E. Gordon, L.S. Rothman, P. Wcislo, C. Hill, J.S. Wilzewski,
           HITRAN Application Programming Interface (HAPI): A comprehensive approach
           to working with spectroscopic data, J. Quant. Spectrosc. Radiat. Transfer 177, 15-30 (2016)      
           DOI: 10.1016/j.jqsrt.2016.03.005

           ATTENTION: This is the core version of the HITRAN Application Programming Interface.
                      For more efficient implementation of the absorption coefficient routine,
                      as well as for new profiles, parameters and other functional,
                      please consider using HAPI2 extension library.
                      HAPI2 package is available at http://github.com/hitranonline/hapi2

Downloading isotopes: 100%|██████████████████████████████████████████████████| 9/9 [01:31<00:00, 10.16s/it]

HITRAN database download complete
Added HITRAN-CO2 database in C:\Users\Tanisha\radis.json
c:\Users\Tanisha\Downloads\RADIS\radis\radis\misc\warning.py:427: HighTemperatureWarning: HITRAN is valid for low temperatures (typically < 700 K). For higher temperatures you may need HITEMP or CDSD. See the 'databank=' parameter
  warnings.warn(WarningType(message))
Calculating Equilibrium Spectrum
Physical Conditions
----------------------------------------
   Tgas                 2000 K
   isotope              1
   mole_fraction        1
   path_length          1 cm
   pressure             1.01325 bar
   self_absorption      True
   species              CO2
   state                X
   wavenum_max          3505.0000 cm-1
   wavenum_min          3500.0000 cm-1
Computation Parameters
----------------------------------------
   Tref                 296 K
   add_at_used          numpy
   broadening_method    voigt
   cutoff               0 cm-1/(#.cm-2)
   dbformat             hitran
   dbpath               C:\Users\Tanisha\.radisdb\hitran\CO2.h5
   diluent              air
   folding_thresh       1e-06
   include_neighbouring_lines  True
   isatom               False
   isneutral            None
   lbfunc               None
   memory_mapping_engine  auto
   neighbour_lines      0 cm-1
   optimization         simple
   parfuncfmt           hapi
   parsum_mode          full summation
   pfsource             default
   potential_lowering   None
   pseudo_continuum_threshold  0
   sparse_ldm           True
   truncation           50 cm-1
   waveunit             cm-1
   wstep                0.01 cm-1
   zero_padding         502
----------------------------------------
0.14s - Spectrum calculated
```
```
--------------------------------------------------------------------------------
CO2 - GEISA 2020 - Fetching database
--------------------------------------------------------------------------------
Attempting to download https://aeris-geisa.ipsl.fr/geisa_files/2020/Lines/line_GEISA2020_asc_gs08_v1.0_co2  
--------------------------------------------------------------------------------
CO2 - GEISA
--------------------------------------------------------------------------------

Download:
- Downloading line_GEISA2020_asc_gs08_v1.0_co2 (1/1)
Downloading: 100%|██████████████████████████████████████████████████████| 128M/128M [00:19<00:00, 7.01MB/s]
Successfully downloaded https://aeris-geisa.ipsl.fr/geisa_files/2020/Lines/line_GEISA2020_asc_gs08_v1.0_co2
Added GEISA-CO2 database in C:\Users\Tanisha\radis.json
```

minouHub reviewed (changes requested):
Quick review from my side.

- [ ] The title framed with '---' is nice. Concerning ExoMol, this title appears *after* the downloading of the states. 
- [ ] Apply linting - [check doc](https://radis.readthedocs.io/en/latest/dev/developer.html#code-style-and-linting)
- [ ] Merge develop to pass the tests. Note I won't merge anything not passing the tests

minouHub commented:
Additional comment. There is an inconsistent behavior if databases are already downloaded and cached.

from radis import SpectrumFactory

```python
sf = SpectrumFactory(3500, 3505,         # cm-1
                     isotope='1',
                     molecule='CO2',
                     verbose=1,    # more for more details
                     )

sf.fetch_databank("hitemp", "2024")
sf.fetch_databank("hitran")
sf.fetch_databank("geisa")
```
output:
```
--------------------------------------------------------------------------------
CO2 - HITEMP 2024 - Downloading and processing 1 chunks for range 3500.0-3505.0 cm⁻¹
--------------------------------------------------------------------------------
- All files already cached.
Processing chunks: 100%|██████████| 1/1 [00:00<00:00, 51.47it/s, from cache]- Combining parsed data from all chunks...
```
I'm expecting something more in these lines:
```
--------------------------------------------------------------------------------
CO2 - HITEMP 2024 - Downloading and processing 1 chunks for range 3500.0-3505.0 cm⁻¹
--------------------------------------------------------------------------------
Download:
- All files already downloaded.
Caching to HDF5/H5 format:
- All files already cached.
--------------------------------------------------------------------------------
CO2 - HITRAN - Downloading database
--------------------------------------------------------------------------------
Download:
- All files already downloaded.
Caching to HDF5/H5 format:
- All files already cached.
--------------------------------------------------------------------------------
CO2 - GEISA
--------------------------------------------------------------------------------
Download:
- All files already downloaded.
Caching to HDF5/H5 format:
- All files already cached.
```

To summarize, if `verbose > 0` and for all databases:

- [ ] Show the subsection "Download" in any case. 
- [ ] Add a comment "All files already downloaded." when it applied
- [ ] Add a subsection "Caching".

We are nearly there!

tanishadahale committed:
Merge branch 'radis:develop' into progress-bar

tanishadahale committed:
Fix linting issues (flake8)

tanishadahale committed:
Standardize cached database output and fix ExoMol header

tanishadahale commented:
Hello @minouHub Thanks for the review!

All points have been addressed:

- ExoMol header: Fixed positioning; title now appears correctly before states download.
- Linting: Resolved
- Merge: Synced with develop branch; all tests passing.
- Cached output: Standardized for HITEMP, HITRAN, and GEISA. Even when cached, output now consistently shows:

  -   Download: “All files already downloaded.”
  -   Caching: “All files already cached.”

tanishadahale committed:
feat: updated progress bar implementation and tests (cleaned)

minouHub commented:
I retried my example 

```python
from radis import SpectrumFactory
sf = SpectrumFactory(3500, 3505,         # cm-1
                     isotope='1',
                     molecule='CO2',
                     verbose=1,    # more for more details
                     )

sf.fetch_databank("hitemp", "2024")
sf.fetch_databank("hitran")
sf.fetch_databank("geisa")
```

```
--------------------------------------------------------------------------------
CO2 - HITEMP 2024 - Downloading and processing 1 chunks for range 3500.0-3505.0 cm⁻¹
--------------------------------------------------------------------------------

Download:
- All files already downloaded.

Caching to HDF5/H5 format:
- All files already cached.
Processing chunks: 100%|██████████| 1/1 [00:00<00:00, 88.90it/s, from cache]- Combining parsed data from all chunks...

--------------------------------------------------------------------------------
CO2 - HITRAN - Downloading database
--------------------------------------------------------------------------------

Download:
- Downloading all isotopes for CO2
Downloading isotopes: 100%|██████████| 9/9 [00:43<00:00,  4.88s/it]

HITRAN database download complete
Added HITRAN-CO2 database in C:\Users\Nicolas Minesi\radis.json
--------------------------------------------------------------------------------
CO2 - GEISA 2020 - Fetching database
--------------------------------------------------------------------------------

Download:
- Attempting to download https://aeris-geisa.ipsl.fr/geisa_files/2020/Lines/line_GEISA2020_asc_gs08_v1.0_co2
--------------------------------------------------------------------------------
CO2 - GEISA
--------------------------------------------------------------------------------

Download:
- Downloading line_GEISA2020_asc_gs08_v1.0_co2 (1/1)
Downloading: 100%|██████████| 128M/128M [00:05<00:00, 23.6MB/s] 

Successfully downloaded https://aeris-geisa.ipsl.fr/geisa_files/2020/Lines/line_GEISA2020_asc_gs08_v1.0_co2

Caching to HDF5/H5 format:
- All files already cached.
Added GEISA-CO2 database in C:\Users\Nicolas Minesi\radis.json
```
Things to do:
- [ ] CO2 GEISA is repeated

tanishadahale committed:
fix: remove duplicate GEISA header in database download

tanishadahale commented:
> Things to do:
> 
> CO2 GEISA is repeated

@minouHub  Fixed!

minouHub commented:
Hello,
Retried. The subsection "Download:" appears twice. (We could have caught it before actually.)
```
--------------------------------------------------------------------------------
CO2 - GEISA 2020 - Fetching database
--------------------------------------------------------------------------------

Download:
- Attempting to download https://aeris-geisa.ipsl.fr/geisa_files/2020/Lines/line_GEISA2020_asc_gs08_v1.0_co2

Download:
- Downloading line_GEISA2020_asc_gs08_v1.0_co2 (1/1)
Downloading: 100%|██████████| 128M/128M [00:10<00:00, 12.8MB/s] 

Successfully downloaded https://aeris-geisa.ipsl.fr/geisa_files/2020/Lines/line_GEISA2020_asc_gs08_v1.0_co2

Caching to HDF5/H5 format:
- All files already cached.
Added GEISA-CO2 database in C:\Users\Nicolas Minesi\radis.json
```

tanishadahale committed:
Fix duplicate 'Download:' section header in database fetching

The 'Download:' section was being printed twice when fetching databases
(GEISA, HITRAN, etc.) - once by the calling function and once by
download_and_parse(). Removed the duplicate from download_and_parse()
since all calling functions already print this header before calling
the method.

tanishadahale committed:
remove duplicate download section

Fix duplicate 'Download:' section header in database fetching

tanishadahale commented:
Done!

```
--------------------------------------------------------------------------------
CO2 - GEISA 2020 - Fetching database
--------------------------------------------------------------------------------

Download:
- Attempting to download https://aeris-geisa.ipsl.fr/geisa_files/2020/Lines/line_GEISA2020_asc_gs08_v1.0_co2
- Downloading line_GEISA2020_asc_gs08_v1.0_co2 (1/1)
Downloading: 100%|█████████████████████████████████████████████████████████████████████████████| 128M/128M [00:23<00:00, 5.79MB/s]

Successfully downloaded https://aeris-geisa.ipsl.fr/geisa_files/2020/Lines/line_GEISA2020_asc_gs08_v1.0_co2

Caching to HDF5/H5 format:
- All files already cached.
Added GEISA-CO2 database in C:\Users\Tanisha\radis.json
```
